### DISCUSSION TITLE
Correct Dagster approach for schedule-less etl?

### DISCUSSION THREAD

**Discussion by jordandakota** [dagster-io/dagster#18478] [Q&A]
Hey All, I'm trying to put together a process using Dagster that we've already accomplished in Airflow, but I'd like to use the most Dagster approved approach possible. Issue is, with assets, multi assets, source assets, asset factories, ops/jobs, I'm not sure which approach is correct, hence this question. I'll lay out the idea:

- We have an app driven approach for listing source assets and defining how to load them (business logic, cron/freshness, etc)
- Connection secrets are stored in a secret manager
- Metadata stored in a postgres instance that has all the tables, their source systems, and their config on how and what to load
- Metadata table has a column that's essentially "next load time"
- Orchestrator should scan this table for new rows where "next load time" > "now", materialize these assets using config
- All of the above should be registered as source assets, but the number of assets each run will obviously differ
- Physical flow is database > local parquet > local duckdb with create or replace/merge into logic as part of the config
- Downstream, there's dbt models that use these source assets complete with tagging and descriptions
- Everything but for the source assets should be set on auto materialize

End goal is, set up the source asset and config it including schedule in our internal app and any downstream models should auto materialize as their upstream's refresh, with every asset being visible in the dagster web server.

My thoughts on the right approach is:

- Use ops and jobs to materialize the source assets into duckdb since our logic is pretty custom and I can't see how to fit them into assets immediately
- Use a source asset factory over a duckdb resource and register all assets by using information schema (how to schedule this?)
- Set up an observable source asset factory over all the above source assets
- Create and integrate dbt as normal and set auto materialize policy at the project level to eager but enable skip on not all parents updated

Have I got this approach correct? What would you change?

**Comment by sryza** [comment]
Instead of a source asset per row in your metadata table, would it make sense to have a single dynamically partitioned asset with a partition per row? Then you could have a sensor that checks that table, creates new partitions as needed, and launches runs for partitions as needed.

**Comment by jordandakota** [comment]
Each row in the metadata table represents a table in some external database though. Essentially, the app allows someone to onboard a new database connection and then config which tables to load. Those loaded tables should be source assets. Does that make sense in a dynamically partitioned asset?
________________________________

**Reply by sryza** [reply]
Yes - it's perfectly common for each partition in a dynamically partitioned to correspond to a table, and for all of them to be sources.

This discussion goes a little more in-depth on this: https://github.com/dagster-io/dagster/discussions/12061.

**Reply by jordandakota** [reply]
Thanks I'll give it a read. Wouldn't even begin to think of a dynamically partitioned asset as separate assets unless prompted.

**Reply by jordandakota** [reply]
Since its a common use case, would you happen to know of an existing example? I've searched around and couldn't find one. Lots of good discussion on use cases though. If not, I'll work through it and start a discussion outlining my steps.
